quantum/audio/audio.c

```diff
@@ -0,0 +1,565 @@
+ * instrument/voice at a time (think: chords). since the number of tones that
+ * can be reproduced depends on the hardware/driver in use: pwm can only
+ * reproduce one tone per output/speaker; DACs can reproduce/mix multiple
+ * when doing additive synthese.
+ *
+ * 'duration' can either be in the beats-per-minute related unit found in
+ * musical_notes.h, OR in ms; keyboards create SONGs with the former, while
+ * the internal state of the audio system does its calculations with the later - ms
+ */
+
+#ifndef AUDIO_TONE_STACKSIZE
+#    define AUDIO_TONE_STACKSIZE 8
+#endif
+uint8_t active_tones                      = 0;       // number of tones pushed onto the stack by audio_play_tone - might be more than the harware is able to reproduce at any single time
+musical_tone_t tones[AUDIO_TONE_STACKSIZE];  // stack of currently active tones
+
+bool playing_melody = false;  // playing a SONG?
+bool playing_note   = false;  // or (possibly multiple simultaneous) tones
+bool state_changed  = false;  // global flag, which is set if anything changes with the active_tones
+
+// melody/SONG related state variables
+float (*notes_pointer)[][2];                // SONG, an array of MUSICAL_NOTEs
+uint16_t notes_count;                       // length of the notes_pointer array
+bool     notes_repeat;                      // PLAY_SONG or PLAY_LOOP?
+uint16_t melody_current_note_duration = 0;  // duration of the currently playing note from the active melody, in ms
+uint8_t  note_tempo      = TEMPO_DEFAULT;  // beats-per-minute
+uint16_t current_note    = 0;              // index into the array at notes_pointer
+bool     note_resting    = false;          // if a short pause was introduced between two notes with the same frequency while playing a melody
+uint16_t last_timestamp  = 0;
+
+#ifdef AUDIO_ENABLE_TONE_MULTIPLEXING
+#    ifndef AUDIO_MAX_SIMULTANEOUS_TONES
+#        define AUDIO_MAX_SIMULTANEOUS_TONES 3
+#    endif
+uint16_t tone_multiplexing_rate        = AUDIO_TONE_MULTIPLEXING_RATE_DEFAULT;
+uint8_t  tone_multiplexing_index_shift = 0;     // offset used on active-tone array access
+#endif
+
+// proviced and used by voices.c
+extern uint8_t  note_timbre;
+extern bool     glissando;
+extern bool     vibrato;
+extern uint16_t voices_timer;
+
+
+#ifndef STARTUP_SONG
+#    define STARTUP_SONG SONG(STARTUP_SOUND)
+#endif
+#ifndef AUDIO_ON_SONG
+#    define AUDIO_ON_SONG SONG(AUDIO_ON_SOUND)
+#endif
+#ifndef AUDIO_OFF_SONG
+#    define AUDIO_OFF_SONG SONG(AUDIO_OFF_SOUND)
+#endif
+float startup_song[][2]   = STARTUP_SONG;
+float audio_on_song[][2]  = AUDIO_ON_SONG;
+float audio_off_song[][2] = AUDIO_OFF_SONG;
+
+static bool    audio_initialized    = false;
+static bool    audio_driver_stopped = true;
+audio_config_t audio_config;
+
+void audio_init() {
+    if (audio_initialized) {
+        return;
+    }
+
+    // Check EEPROM
+    // TODO: avr defines?
+#if defined(STM32_EEPROM_ENABLE) || defined(PROTOCOL_ARM_ATSAM) || defined(EEPROM_SIZE)
```

Comment:
This should be fine: 
```suggestion
#ifdef EEPROM_ENABLE
```

Specifically, it's handled for AVR automatically:
https://github.com/qmk/qmk_firmware/blob/28e5969f780ec4a29ee27f55af60fab88369d1c4/common_features.mk#L73-L126

Reply:
thanks! that block was "inherited" from before the refactoring